plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.jvm.abi
+
+import org.jetbrains.kotlin.backend.common.output.OutputFile
+import org.jetbrains.kotlin.backend.common.output.OutputFileCollection
+import org.jetbrains.kotlin.backend.common.output.SimpleOutputBinaryFile
+import org.jetbrains.kotlin.cli.common.messages.MessageCollector
+import org.jetbrains.kotlin.cli.common.output.writeAllTo
+import org.jetbrains.kotlin.cli.jvm.compiler.CompileEnvironmentUtil
+import org.jetbrains.kotlin.codegen.ClassFileFactory
+import org.jetbrains.kotlin.codegen.extensions.ClassFileFactoryFinalizerExtension
+import org.jetbrains.kotlin.load.java.JvmAnnotationNames
+import org.jetbrains.org.objectweb.asm.*
+import org.jetbrains.org.objectweb.asm.commons.Method
+import org.jetbrains.org.objectweb.asm.tree.*
+import java.io.File
+
+class JvmAbiOutputExtension(
+    private val outputPath: File,
+    private val abiClassInfos: Map<String, AbiClassInfo>,
+) : ClassFileFactoryFinalizerExtension {
+    override fun finalizeClassFactory(factory: ClassFileFactory) {
+        // We need to wait until the end to produce any output in order to strip classes
+        // from the InnerClasses attributes.
+        val outputFiles = AbiOutputFiles(abiClassInfos, factory)
+        if (outputPath.extension == "jar") {
+            // We don't include the runtime or main class in interface jars and always reset time stamps.
+            CompileEnvironmentUtil.writeToJar(outputPath, false, true, true, null, outputFiles, MessageCollector.NONE)
```

Comment:
...I completely missed that. Changed the code to use the same `MessageCollector` as the rest of the compiler. That should be fine, since the only diagnostic that `writeToJar` can report is an invalid jar path error.